### PR TITLE
[#IOPID-376] The LolliPoP middleware can complete without an user session

### PR DESCRIPTION
**Pull request by BurnedMarshal**
#### List of Changes

New `withOptionalUserFromRequest` to retrieve the User if present
The Lollipop middleware verify the CF and AssertionRef from the request and the LCParams if the session is missing
add some validation on edge case

#### Motivation and Context

LV needs to use LolliPoP without an active user session. The Lollipop middleware now can be used even if the session token is not provided.

#### How Has This Been Tested?

unit test

#### Types of changes

- [X] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/utils/lollipop.ts [hunk 1]
```diff
@@ -166,25 +218,41 @@ export const extractLollipopLocalsFromLollipopHeaders = (
...
     TE.map(
-      (lcParams) =>
+      ({ lcParams }) =>
         ({
           ["x-pagopa-lollipop-assertion-ref"]: lcParams.assertion_ref,
           ["x-pagopa-lollipop-assertion-type"]: lcParams.assertion_type,
           ["x-pagopa-lollipop-auth-jwt"]: lcParams.lc_authentication_bearer,
           ["x-pagopa-lollipop-public-key"]: lcParams.pub_key,
-          ["x-pagopa-lollipop-user-id"]: fiscalCode,
+          // It's possible to improve security by verify that the fiscal code from
```
Comment (arcogabbo) [inline src/utils/lollipop.ts:253]:
```suggestion
          // It's possible to improve security by verifying that the fiscal code from
          // the authorization is equal to the one from the lollipop function
```

**Review by arcogabbo — APPROVED** [review]
LGTM, just some minor comment

File: src/adapters/__tests__/pnFetch.test.ts [hunk 2]
```diff
@@ -146,10 +150,8 @@ describe("getThirdPartyMessageDetails", () => {
       "x-pagopa-cx-taxid": aFiscalCode,
       signature:
         "sig1=:hNojB+wWw4A7SYF3qK1S01Y4UP5i2JZFYa2WOlMB4Np5iWmJSO0bDe2hrYRbcIWqVAFjuuCBRsB7lYQJkzbb6g==:",
-      "signature-input":
-        'sig1=("x-pagopa-lollipop-original-method" "x-pagopa-lollipop-original-url"); created=1618884475; keyid="test-key-rsa-pss"',
-      "x-pagopa-lollipop-assertion-ref":
-        "sha256-6LvipIvFuhyorHpUqK3HjySC5Y6gshXHFBhU9EJ4DoM=",
+      "signature-input": aSignatureInput,
```
Comment (arcogabbo) [inline src/adapters/__tests__/pnFetch.test.ts:153]: also, to reduce code duplication here we could create a constant with the object provided to this upper function `toHaveBeenCalledWith`, ready to be used across all the file.

File: src/utils/lollipop.ts (hunk 1 above)
Comment (BurnedMarshal) [reply src/utils/lollipop.ts]: Fixed in [06a2ed3](https://github.com/pagopa/io-backend/pull/1033/commits/06a2ed3c4f90691ff80f51bb2b091ad054e62149)

File: src/adapters/__tests__/pnFetch.test.ts (hunk 2 above)
Comment (BurnedMarshal) [reply src/adapters/__tests__/pnFetch.test.ts]: Fixed in [e919283](https://github.com/pagopa/io-backend/pull/1033/commits/e9192839d7cddfb5a992b8763defcb26ba16db65)

**Pull request merged**